### PR TITLE
OCI Handler Unit Tests

### PR DESCRIPTION
**Pull request by Shaheer-Ahmd**
# Add Unit Tests for OCI Handler Functions

This PR introduces new test cases for the `ocihandler` package, particularly focusing on the `checkBuilderVersion`, `getPullSecret`, and `constructTempConfig` functions. The changes aim to improve test coverage by verifying different scenarios such as handling invalid annotations, secrets, and configurations. The tests include edge cases like missing annotations, invalid secret types, and incorrect configurations, ensuring that the functions behave as expected under various conditions.

The test cases for `checkBuilderVersion` validate the handling of builder version discrepancies, while tests for `getPullSecret` ensure that secret retrieval behaves correctly for valid and invalid secrets. Additionally, the tests for `constructTempConfig` check various annotation formats and handle edge cases like empty fields or missing values. These improvements enhance the robustness of the code by thoroughly testing corner cases and ensuring proper error handling.

an effort towards #3835 

## How to use

Reviewers can validate this PR by running the tests using the following command:

```
go test -v ./pkg/operators/oci-handler/
```

Ensure that all tests pass without any errors, and verify that the test results cover a range of scenarios, including error handling and edge cases.

## Testing done

The testing was performed by executing the following command:

```
go test -v ./pkg/operators/oci-handler/
```

The output included successful validation for various test cases, covering different scenarios such as valid secrets, invalid annotations, and handling of edge cases in configuration parsing. The log output was checked for expected log messages, ensuring that the correct error handling and logging behavior was exhibited for each case.

### CODE REVIEW

File: pkg/operators/oci-handler/oci_test.go [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+		// 	input:   "source:foo=bar=baz",
+		// 	wantErr: true,
+		// },
+		// {
+		// 	name:    "Invalid multiple dots",
+		// 	input:   "source.field.foo:foo=bar",
+		// 	wantErr: true,
+		// },
+		{
+			name:    "Empty annotation key",
```
Comment (Shaheer-Ahmd) [inline pkg/operators/oci-handler/oci_test.go:270]: Reviewer pls confirm this

File: pkg/operators/oci-handler/oci_test.go [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+		{"v1.0.0", "Different version warn", "0.9.0", "This gadget was built with ig 0.9.0 and it's being run with v1.0.0. Gadget could be incompatible"},
+		{"v1.0.0", "Matching version no log", "1.0.0", ""},
+	}
+
+	lg := logrus.StandardLogger()
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// version.Version = func() *semver.Version { return tc.currentVersion }
+			logBuffer.Reset()
```
Comment (Shaheer-Ahmd) [inline pkg/operators/oci-handler/oci_test.go:53]: ignore these comments pls (will remove these dead comments in the final commit once the PR is reviewed)

File: pkg/operators/oci-handler/oci_test.go [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+							"": map[string]any{
+								"annotations": map[string]any{
+									"foo": "bar",
+								},
+							},
+						},
+					},
+				},
+			},
+			wantLength: 2,
```
Comment (burak-ok) [inline pkg/operators/oci-handler/oci_test.go:331]: I think for the 3 test cases I adjusted we also want to return an error for them. Could you fix the logic in `oci.go` for them in a separate commit?
cc @flyth 
```suggestion
		{
			name:    "Empty annotation key",
			input:   "source:=bar",
			wantErr: true,
		},
		{
			name:    "Empty annotation value",
			input:   "source:foo=",
			wantErr: false,
			wantConfig: map[string]any{
				"datasources": map[string]any{
					"source": map[string]any{
						"annotations": map[string]any{
							"foo": "",
						},
					},
				},
			},
			wantLength: 1,
		},
		{
			name:    "Empty subject",
			input:   ":foo=bar",
			wantErr: true,
		},
		{
			name:    "Empty field name",
			input:   "source.:foo=bar",
			wantErr: true,
```

File: pkg/operators/oci-handler/oci_test.go [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+	}
+}
+
+func TestConstructTempConfig(t *testing.T) {
+	tests := []struct {
+		name       string
+		input      string
+		wantErr    bool
+		wantConfig map[string]any
+		wantLength int
```
Comment (burak-ok) [inline pkg/operators/oci-handler/oci_test.go:207]: I am not so sure about the naming. Its not the length we care about, but the type of annotation we get back from the function. Its either a `DataSource` annotation or a `Field` annotation

File: pkg/operators/oci-handler/oci_test.go [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+		},
+		// TODO: Shouldn't these be invalid?
+		{
+			name:    "Invalid multiple colons",
+			input:   "source:field:foo=bar",
+			wantErr: true,
+		},
+		{
+			name:    "Invalid multiple equal signs",
+			input:   "source:foo=bar=baz",
```
Comment (burak-ok) [inline pkg/operators/oci-handler/oci_test.go:261]: This should be valid. There might be cases where the value might be base64 encoded and therefore the value could have multiple `=` equal signs

File: pkg/operators/oci-handler/oci_test.go [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+			wantErr: true,
+		},
+		{
+			name:    "Invalid multiple dots",
+			input:   "source.field.foo:foo=bar",
```
Comment (burak-ok) [inline pkg/operators/oci-handler/oci_test.go:266]: I don't know why I forgot, but this should be also successful (just subfields or subsubfields, ...)

File: pkg/operators/oci-handler/oci.go [hunk 7]
```diff
@@ -237,6 +233,69 @@ func checkBuilderVersion(manifest *v1.Manifest, logger logger.Logger) {
...
+	// empty subject
+	if slices.Contains(annInfo, "") {
+		return nil, 0, fmt.Errorf("invalid annotation %q", ann)
+	}
+
+	annotation := strings.SplitN(annInfo[1], "=", 2)
+	if len(annotation) != 2 {
+		return nil, 0, fmt.Errorf("invalid annotation %q", ann)
+	}
+	// empty key or value
```
Comment (burak-ok) [inline pkg/operators/oci-handler/oci.go:254]: We might want to allow empty values IMO

File: pkg/operators/oci-handler/oci_test.go [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+			name:    "Empty subject",
+			input:   ":foo=bar",
+			wantErr: true,
+		},
+		{
+			name:    "Empty field name",
+			input:   "source.:foo=bar",
+			wantErr: true,
+		},
+		{
```
Comment (burak-ok) [inline pkg/operators/oci-handler/oci_test.go:328]:
```suggestion
		{
			name:    "Empty subfield name",
			input:   "source.test..bar:foo=bar",
			wantErr: true,
		},
		{
```

I added a new testcase, which should fail, but is accepted by the proposed code changes in the first commit

File: pkg/operators/oci-handler/oci.go (hunk 7 above)
Comment (Shaheer-Ahmd) [reply pkg/operators/oci-handler/oci.go:254]: ok. 
would you like me to make the error messages more descriptive. right now, for all the cases, the error message is "invalid annotation".

Comment (burak-ok) [reply pkg/operators/oci-handler/oci.go:254]: That would be even better :)

File: pkg/operators/oci-handler/oci_test.go [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+			},
+			isField: true,
+		},
+		{
+			name:    "Invalid multiple colons",
+			input:   "source:field:foo=bar",
+			wantErr: true,
+		},
+		{
+			name:    "multiple equal signs (base64 encoded)",
```
Comment (burak-ok) [inline pkg/operators/oci-handler/oci_test.go:262]:
```suggestion
			name:    "allowing equal signs in value (e.g. base64 encoded)",
```

File: pkg/operators/oci-handler/oci_test.go [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+					"source": map[string]any{
+						"annotations": map[string]any{
+							"foo": "bar=baz",
+						},
+					},
+				},
+			},
+		},
+		{
+			name:    "multiple colon signs in value (base64 encoded)",
```
Comment (burak-ok) [inline pkg/operators/oci-handler/oci_test.go:276]:
```suggestion
			name:    "allowing colon signs in value",
```

**Review by burak-ok — APPROVED** [review]
After the 2 small changes this is good to go.

Thanks :)
